src/cpp/runner/main.cpp

```diff
@@ -9,26 +10,144 @@
 
 using namespace std;
 
-/*GMMOutput readInputData() {
+GMMInput read_input_data(const string& input_file, const bool replicate_point)
+{
+    GMMInput input;
+
+    // Read instance
+    read_gmm_instance(input_file, &input.d, &input.k, &input.n,
+        input.alphas, input.means, input.icf, input.x, input.wishart, replicate_point);
+
+    return input;
+}
+
+typedef  void (ITest<GMMInput, GMMOutput>::* test_member_function) (int);
+
+inline void call_member_function(unique_ptr<ITest<GMMInput, GMMOutput>>& ptr_to_object, const test_member_function ptr_to_member, const int arg1) {
+    (*(ptr_to_object).*(ptr_to_member))(arg1);
+}
+
+double measure_shortest_time(const double minimum_measurable_time, const int nruns, const double time_limit, unique_ptr<ITest<GMMInput, GMMOutput>>& test, const test_member_function func)
+{
+    vector<double> samples;
+    double total_time = 0;
+    auto repeats = 1;
+    for (;; repeats *= 2)
+    {
+        auto t1 = high_resolution_clock::now();
+        call_member_function(test, func, repeats);
+        auto t2 = high_resolution_clock::now();
+        const auto current_run_time = duration_cast<duration<double>>(t2 - t1).count();
+        if (current_run_time > minimum_measurable_time)
+        {
+            samples.push_back(current_run_time / repeats);
+            total_time += current_run_time;
+            break;
+        }
+    }
+
+    for (auto run = 1; (run < nruns) && (total_time < time_limit); run++)
+    {
+        auto t1 = high_resolution_clock::now();
+        call_member_function(test, func, repeats);
+        auto t2 = high_resolution_clock::now();
+        const auto current_run_time = duration_cast<duration<double>>(t2 - t1).count();
+        samples.push_back(current_run_time / repeats);
+        total_time += current_run_time;
+    }
+
+    const auto min_sample = *std::min_element(std::begin(samples), std::end(samples));
```

Comment:
We are computing the minimum sample and and the total time. Both of these can be computed by folding, so we don't really need a vector here.